web/js/widgets/06_switcher.js

```diff
@@ -0,0 +1,152 @@
+function Switcher(grid, uuid) {
+  this.color = '#000000'
+  this.fontSize = 14
+  this.value = 0
+  this.__switcher = true
+  this.grid = grid
+
+  var el = document.createElement('div')
+  Widget.call(this, grid, uuid, el)
+
+  var canvas = document.createElement('canvas')
+  canvas.style.width = '100%'
+  canvas.style.height = '100%'
+
+  this.input = document.createElement('input')
+  this.input.type = 'text'
+  this.input.style.all = 'revert'
+  this.input.style.height = '100%'
+  this.input.style.borderWidth = '2px'
+  this.input.style.boxSizing = 'border-box'
+  this.input.style.borderColor = '#cccccc'
+  this.input.style.position = 'absolute'
+
+  el.appendChild(canvas)
+  el.appendChild(this.input)
+
+  this.w = 4
+  this.h = 1
+
+  this.canvas = ge1doot.canvas(canvas)
+  this.canvas.resize = this.draw.bind(this)
+  this.canvas.pointer.up = function () {
+    var bounds = this.el.getBoundingClientRect()
+    const width = bounds.width
+    const height = bounds.height
+    const part = width >= 3 * height
+      ? height
+      : this.canvas.width / 4
+    const inputWidth = width >= 3 * height
+      ? (width - 2 * height)
+      : width / 2
+    if (this.canvas.pointer.x < part) {
+      this.value -= 1
+      if (this.value < this.min)
+        this.value = this.min
+    } else if (this.canvas.pointer.x > part + inputWidth) {
+      this.value += 1
+      if (this.value > this.max)
+        this.value = this.max
+    } else {
+      return
+    }
+
+    this.sendEvent('changed', { value: this.value })
+    this.draw()
+  }.bind(this)
+
+  this.input.addEventListener(
+    'change',
+    function () {
+      const val = parseInt(this.input.value, 10)
+      if (val < this.min)
+        val = this.min
+      else if (val > min(this.max, this.grid.tabs.length))
+        val = min(this.max, this.grid.tabs.length)
+      this.value = val
+      this.draw()
+      this.sendEvent('changed', { value: this.value })
+    }.bind(this)
+  )
+}
+
+Widget.createSubclass(Switcher, {
+  fontSize: new Prop(Number),
+  color: new Prop(String).setIsColor(),
+  value: new Prop(Number),
+  min: new Prop(Number),
+  max: new Prop(Number)
```

Comment:
K čemu je to min a max? Jakože některé taby nepůjdou vybrat pomocí tohodle switcheru?

Reply:
Ano, kdybych chtěl vytvořit skryté taby